---
title: "Seurat Workflow and Cluster Assignment for Batch 6"
output: html_document
---

```{R}
library(Seurat)
library(cowplot)
library(ggplot2)
library(ggsci)
library(dplyr)
library(psych)
library(pheatmap)
library(harmony)
#library(clusterProfiler)
#library(org.Hs.eg.db)
library(DOSE)
library(GOSemSim)
library(enrichplot)
library(glmpca)
library(SeuratWrappers)
```

```{R}

#Batch2.data <- Read10X(data.dir = "Files/cellranger_matrices/cellranger_aggr_matrices_old/filtered_feature_bc_matrix")

```


```{R}

#dim(Batch2.data)

```

Loading in the single cell data

```{r}
BN0329.data <- Read10X(data.dir = "Files/cellranger_matrices/BN0329")
BN0347.data <- Read10X(data.dir = "Files/cellranger_matrices/BN0347")
BN0348.data <- Read10X(data.dir = "Files/cellranger_matrices/BN0348")
BN0464.data <- Read10X(data.dir = "Files/cellranger_matrices/BN0464")
```






checking dimensions to check all samples have the same set of genes

```{r}
dim(BN0329.data)
dim(BN0347.data)
dim(BN0348.data)
dim(BN0464.data)
```


```{r}
Batch6_MTG <- CreateSeuratObject(counts = cbind(BN0329.data,
                                           BN0347.data,
                                           BN0348.data,
                                           BN0464.data),
                            project = "Batch6_MTG",
                            min.cells = 3)
```


Let's now annotate columns (cells) with meta data :

1) sample ID
```{r}
Batch6_MTG@meta.data$sample_ID <- c(rep("BN0329", ncol(BN0329.data)),
                               rep("BN0347", ncol(BN0347.data)),
                               rep("BN0348", ncol(BN0348.data)),
                               rep("BN0464", ncol(BN0464.data))
)
```

2) case : Healthy Control (HC) or Parkinson Disease (PD)
```{r}
Batch6_MTG@meta.data$case <- c(rep("PD", ncol(BN0329.data)),
                          rep("HC", ncol(BN0347.data)),
                          rep("PD", ncol(BN0348.data)),
                          rep("ILB", ncol(BN0464.data))
)
```

3) Expermient (in this case experiment and sample_ID are identical.)
```{r}
Batch6_MTG@meta.data$experiment <- c(rep("BN0329", ncol(BN0329.data)),
                                rep("BN0347", ncol(BN0347.data)),
                                rep("BN0348", ncol(BN0348.data)),
                                rep("BN0464", ncol(BN0464.data))
)
```
 
4) One VS all annorations :
BN0329 vs rest
```{r}
Batch6_MTG@meta.data$BN009_vs_rest <- c(rep("BN0329", ncol(BN0329.data)),
                                    rep("rest", ncol(BN0347.data)),
                                    rep("rest", ncol(BN0348.data)),
                                    rep("rest", ncol(BN0464.data)))
```

BN0347 vs rest

```{r}
Batch6_MTG@meta.data$BN009_vs_rest <- c(rep("rest", ncol(BN0329.data)),
                                    rep("BN0347", ncol(BN0347.data)),
                                    rep("rest", ncol(BN0348.data)),
                                    rep("rest", ncol(BN0464.data)))
```

BN0348 vs rest
```{r}
Batch6_MTG@meta.data$BN009_vs_rest <- c(rep("rest", ncol(BN0329.data)),
                                    rep("rest", ncol(BN0347.data)),
                                    rep("BN0348", ncol(BN0348.data)),
                                    rep("rest", ncol(BN0464.data)))
```

BN0464 vs rest
```{r}
Batch6_MTG@meta.data$BN009_vs_rest <- c(rep("rest", ncol(BN0329.data)),
                                    rep("rest", ncol(BN0347.data)),
                                    rep("rest", ncol(BN0348.data)),
                                    rep("BN0464", ncol(BN0464.data)))
```



Now let's see the cell the counts for each group:
```{r}
table(Batch6_MTG$sample_ID)
table(Batch6_MTG$case)
table(Batch6_MTG$experiment)
```


 Quality Checking : 

 Seurat allows you to easily explore QC metrics and filter cells based on any user-defined criteria.
 A few QC metrics commonly used by the community include : 
 The number of unique genes detected in each cell.
 Low-quality cells or empty droplets will often have very few genes
 Cell doublets or multiplets may exhibit an aberrantly high gene count
 Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)
 The percentage of reads that map to the mitochondrial genome
 Low-quality / dying cells often exhibit extensive mitochondrial contamination
 We calculate mitochondrial QC metrics with the PercentageFeatureSet() function, which calculates the percentage of counts originating from a set of features
 We use the set of all genes starting with MT- as a set of mitochondrial genes

```{r}
Batch6_MTG[["percent.mt"]] <- PercentageFeatureSet(Batch6_MTG, pattern = "^MT-")

VlnPlot(Batch6_MTG, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

 An object of class Seurat 
 25381 features across 6549 samples within 1 assay 
 Active assay: RNA (25381 features)
 
```{R}
library(stats)
nfeature_RNA <- Batch6_MTG@meta.data$nFeature_RNA
mean(nfeature_RNA)
MAD <- mad(nfeature_RNA, center = median(nfeature_RNA))
threeMAD <- (MAD*3)+median(Batch6_MTG@meta.data$nFeature_RNA)

```

Filtering low quality cells:
```{r}
Batch6_MTG <- subset(Batch6_MTG, subset = nFeature_RNA > 200 & nfeature_RNA < threeMAD | percent.mt < 5)
```

Now let's recheck the cell the counts for each group:
```{r}
table(Batch6_MTG$sample_ID)
table(Batch6_MTG$case)
table(Batch6_MTG$experiment)
```


Log Normalizing data: 
```{r}
Batch6_MTG <- NormalizeData(Batch6_MTG, normalization.method = "LogNormalize", scale.factor = 10000)
```


Finding and plotting 2000 most variable features

```{R}
Batch6_MTG <- FindVariableFeatures(Batch6_MTG, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(Batch6_MTG), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(Batch6_MTG)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
```

```{R}
plot1
```

```{R}
plot2
```

Run the standard workflow for visualization and clustering :

save list of all genes
```{r}
all.genes <- rownames(Batch6_MTG)
```

center and scale data
```{r}
Batch6_MTG <- ScaleData(Batch6_MTG, features = all.genes, verbose = FALSE)
```

Determining "Dimensionality" of the dataset





finding the top 30 principal components for cells
```{r}
Batch6_MTG <- RunGLMPCA(Batch6_MTG, features=Batch6_MTG@assays$RNA@var.features, L = 30)
```

see contribution of genes to construct each of these principal components.
```{r}
VizDimLoadings(Batch6_MTG, dims = 1:2, reduction = "pca")
```


plot cells in the first two principal components colored by case: 
```{r}
DimPlot(object = Batch6_MTG, reduction = "pca", pt.size = .1, group.by = "case")
```

Coordinate of cells in PC 1 characterized by case: 
```{r}
VlnPlot(object = Batch6_MTG, features = "PC_1", group.by = "case",  pt.size = .1)
```

Run Harmony
```{r}
Batch6_MTG <- RunHarmony(Batch6_MTG, group.by.vars = c("sample_ID","case"), reduction = "glmpca", plot_convergence = TRUE, theta = c(1,1))
```


cells in harmony axis
```{r}
harmony_embeddings <- Embeddings(Batch6_MTG, 'harmony')
harmony_embeddings[1:5, 1:5]
```

cells in harmony axis
```{r}
DimPlot(object = Batch6_MTG, reduction = "harmony", pt.size = .1, group.by = "case")
```
cells in harmony 1 axis
```{r}
VlnPlot(object = Batch6_MTG, features = "harmony_1", group.by = "case",  pt.size = .1)
```

 An object of class Seurat 
 25381 features across 6549 samples within 1 assay 
 Active assay: RNA (25381 features)
 2 dimensional reductions calculated: pca, harmony

Determing the dimensionality of the dataset

```{R}
Batch6_MTG <- JackStraw(Batch6_MTG, num.replicate = 100)
Batch6_MTG <- ScoreJackStraw(Batch6_MTG, dims = 1:20)
JackStrawPlot(Batch6_MTG, dims = 1:20)



```


```{R}
ElbowPlot(Batch6_MTG, ndims = 30, reduction = "harmony")

```
10 looks like a suitable cutoff based on the elbow plot, but the JackStraw plot looks very strange

Finding Clusters of cells:
```{r}
Batch6_MTG <- FindNeighbors(Batch6_MTG, reduction = "harmony", dims = 1:18)
Batch6_MTG <- FindClusters(Batch6_MTG, resolution = 0.5, algorithm = 4)
```

run Umap based on top 10 harmony axis: 

```{r}
Batch6_MTG <- RunUMAP(Batch6_MTG, reduction = "harmony", dims = 1:18)
```

```{r}
DimPlot(Batch6_MTG, label = TRUE, repel = TRUE, pt.size = 0, label.size = 2)

UMAPclusters <- DimPlot(Batch6_MTG, label = TRUE, repel = TRUE, pt.size = 0, label.size = 2.5) + 
        theme(axis.text = element_text(size=8),
              axis.title = element_text(size = 12),
              legend.text = element_text(size = 8),
              title = element_text(size = 12),
              legend.key.size = unit(0.4,"cm"))

ggsave(UMAPclusters, filename = "Figures/Unassigned_GLMAPCA_UMAPclusters_scRNA_seq_Batch6_MTG.pdf", device = "pdf", width = 6, height = 4, units = "in")
```


```{R}
UMAPclusters <- DimPlot(Batch6_MTG, label = TRUE, repel = TRUE, pt.size = 0, label.size = 2.5, split.by = "case") + 
        theme(axis.text = element_text(size=8),
              axis.title = element_text(size = 12),
              legend.text = element_text(size = 8),
              title = element_text(size = 12),
              legend.key.size = unit(0.4,"cm"))

ggsave(UMAPclusters, filename = "Figures/Unassigned_Split_GLMPCA_UMAPclusters_scRNA_seq_Batch6_MTG.pdf", device = "pdf", width = 12, height = 4, units = "in")
```

run tsne based on top 10 harmony axis: 

```{r}
Batch6_MTG <- RunTSNE(Batch6_MTG, reduction = "harmony", dims = 1:18)
```

plot umap: 
```{r}
DimPlot(Batch6_MTG, reduction = "umap", label = TRUE,pt.size = 0.01)
```

```{r}
DimPlot(Batch6_MTG, reduction = "umap", group.by = "case",pt.size = 0.1)
```

```{r}
DimPlot(Batch6_MTG, reduction = "umap", group.by = "experiment",pt.size = 0.1)
```

TSNE Plot cells colored by clusters and grouped by case. 
```{r}
DimPlot(Batch6_MTG, reduction = "tsne", split.by = "case", label = TRUE, ncol = 1)
```

Find markers for every cluster compared to all remaining cells, report only the positive ones
```{r}
Batch6_MTG.markers <- FindAllMarkers(Batch6_MTG, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
```


Save single-cell elements. 
```{r}
#saveRDS(AllMB,"AllMB.rds")
```

```{r}
#AllMB=readRDS("AllMB.rds")
```

```{r}
#saveRDS(AllMB.markers,"AllMB.markers.rds")
```

```{r}
#AllMB.markers=readRDS("AllMB.markers.rds")
#write.csv(AllMB.markers,"AllMB_markers.csv")
```

```{r}
unique(Batch6_MTG.markers$cluster)
```

Top marker genes for each clusters: 
```{r}
Batch6_MTG.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)
write.table(Batch6_MTG.markers, file = "Files/AllMarkers.txt", col.names = TRUE, sep = "\t", quote = FALSE)
top10Markers <- Batch6_MTG.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
write.table(top10Markers, file = "Files/top10Markers.txt", col.names = TRUE, sep = "\t", quote = FALSE)
```

Heatmap for some marker genes: 
```{r}
features <- unique(top10Markers$gene)
DoHeatmap(Batch6_MTG, features = features, size = 2, draw.lines = FALSE, angle = 45,
          hjust = 0.2) + theme(axis.text.y = element_text(size = 5))
```
#ignore actual genes, this is just a reminder of what a FeaturePlot is 

Plot these genes in the cluster plot:
```{r}
FeaturePlot(Batch6_MTG, features = c("ENO2", "RBFOX3","SLC17A6", "SLC17A7","AQP4", "GFAP"))
```

```{r}
#saveRDS(AllMB,"AllMB_Unassigned.rds")
```


 
 #########################################################
 
```{R}
#AllMB=readRDS("AllMB_Assigned.rds")
```
 
Testing T-cell/immune cell markers




```{r}
VlnPlot(Batch6_MTG, features = c("CD96"),pt.size = 0)
```
NK Cell Markers

```{r}
VlnPlot(Batch6_MTG, features = c("IL7R"),pt.size = 0)
```


CD16

```{r}
VlnPlot(Batch6_MTG, features = c("FCGR3A"),pt.size = 0)
```

CD56

```{r}
VlnPlot(Batch6_MTG, features = c("NCAM1"),pt.size = 0)
```

CD94


```{r}
VlnPlot(Batch6_MTG, features = c("KLRD1"),pt.size = 0)
```
NKG2A

```{r}
VlnPlot(Batch6_MTG, features = c("KLRC1"),pt.size = 0)
```
```{r}
VlnPlot(Batch6_MTG, features = c("CD244"),pt.size = 0)
```

resting T cell marker?


```{r}
VlnPlot(Batch6_MTG, features = c("CD8A"),pt.size = 0)
```

```{r}
TKillerCellMarkerVlnPlot <- VlnPlot(AllMB, features = c("CD8A","CD226","CD244","CD96","IL7R","KLRD1","NCAM1","FCGR3A","KLRC1"),pt.size = 0, ncol = 1) 

ggsave(TKillerCellMarkerVlnPlot,filename = "TKillerCellMarkerVlnPlot.pdf", width = 20, height = 20)            


```


##################################################################

T lymphocyte markers

```{r}
VlnPlot(Batch6_MTG, features = c("CD69"),pt.size = 0)
```



```{r}
VlnPlot(Batch6_MTG, features = c("TFRC"),pt.size = 0)
```



```{r}
VlnPlot(Batch6_MTG, features = c("IL2RA"),pt.size = 0)
```



```{r}
VlnPlot(Batch6_MTG, features = c("CTLA4"),pt.size = 0)
```



```{r}
VlnPlot(Batch6_MTG, features = c("FOXP3"),pt.size = 0)
```

```{r}
VlnPlot(Batch6_MTG, features = c("TRAC"),pt.size = 0)
```


###################################################################

 Neuron  = ENO2, RBFOX3
 Glutamatergic neurons = SLC17A6, SLC17A7
 GABAergic neurons = SLC32A1, GAD1, GAD2
 Dopaminergic neurons = TH, SLC6A3, SCL18A2
 Astrocytes  = AQP4, GFAP
 Oligodendrocytes  =  PLP1, MBP
 OPCs  =  VCAN, BCAN,
 Microglia = CX3CR1, P2RY12
 Endothelial cells = FLT1, CLDN5
 

 Glu_Neurons = 0,1,11,20
 GABA_Neurons = 5,6,9,10,(14),23,24
 Astrocytes = 7,21
 oligo = 3,(4),22
 OPCs= 14
 Microglia = 12
 novel = 2,8,13,15,16,19,20,22,23,24,25 - all likely neurons

Find cell-types by plotting marker genes.

Neuron Markers

```{r}
VlnPlot(Batch6_MTG, features = c("ENO2"),pt.size = 0)
```
Neurons: 1,4,5,6,7,9,11,12,13,14,15,16,19,20,21


```{R}
VlnPlot(Batch6_MTG, features = c("RBFOX3"),pt.size = 0)
```

Glutamatergic neuron markers

```{R}
VlnPlot(Batch6_MTG, features = c("SLC17A6"),pt.size = 0)
```
```{R}
VlnPlot(Batch6_MTG, features = c("SLC17A7"),pt.size = 0)
```
0,1,11,20 - Glu?

GABAergic markers



```{R}
VlnPlot(Batch6_MTG, features = c("SLC32A1"),pt.size = 0)
```
```{R}
VlnPlot(Batch6_MTG, features = c("GAD1"),pt.size = 0)
```
```{R}
VlnPlot(Batch6_MTG, features = c("GAD2"),pt.size = 0)
```

5,6,9,10,(14),23,24 - GABA?

weaker evidence for 10

Astrocyte markers
```{R}
VlnPlot(Batch6_MTG, features = c("AQP4"),pt.size = 0)
```
```{R}
VlnPlot(Batch6_MTG, features = c("GFAP"),pt.size = 0)
```
7,(17,18),21 - Astrocytes

Oligodendrocyte markers


```{R}
VlnPlot(Batch6_MTG, features = c("PLP1"),pt.size = 0)
```
```{R}
VlnPlot(Batch6_MTG, features = c("MBP"),pt.size = 0)
```


```{R}
VlnPlot(Batch6_MTG, features = c("MOG"),pt.size = 0)
```

3,(4),22 - oligodendrocytes - strong



OPC markers

```{R}
VlnPlot(Batch6_MTG, features = c("VCAN"),pt.size = 0)
```
```{R}
VlnPlot(Batch6_MTG, features = c("BCAN"),pt.size = 0)
```

(7),14,(21) - OPCs

Microglia

```{R}
VlnPlot(Batch6_MTG, features = c("CX3CR1"),pt.size = 0)
```
```{R}
VlnPlot(Batch6_MTG, features = c("P2RY12"),pt.size = 0)
```

12 - Microglia

Endothelial Cells

```{R}
VlnPlot(Batch6_MTG, features = c("FLT1"),pt.size = 0)
```
```{R}
VlnPlot(Batch6_MTG, features = c("CLDN5"),pt.size = 0)
```
17,18 - Endothelial cells



Reading in the table for panglaodb to more systematically go through the known types of brain cell and assess their presence in the dataset.

```{R}
PanglaoDB <- read.delim("~/Documents/Projects/Misc. Files/PanglaoDB_markers_27_Mar_2020.tsv", stringsAsFactors = FALSE)

PanglaoDBBrain <- PanglaoDB[PanglaoDB$organ %in% "Brain",]
```

Let's see how many unique cell types for the brain the database contains

```{R}
unique(PanglaoDBBrain$cell.type)

sum(is.na(PanglaoDBBrain$cell.type))
```
Reminder of the unnassigned clusters:

novel = 7,12,14,15,19,21 - all neurons

Looking at top 5 markers for each cell type

Making function and testing

```{R}
GeneSet <- rownames(GetAssayData(object = Batch6_MTG, assay.type = "RNA", slot = "counts"))

#filtering PanglaoDB for genes present in the scRNA-seq dataset

PanglaoDBBrainfilter <- PanglaoDBBrain[PanglaoDBBrain$official.gene.symbol %in% GeneSet,]

markervlnplots <- function(x) {
  VlnPlot(Batch6_MTG, features = x,pt.size = 0)
  
}

markervlnplots("RELN")

```



Running function for Tancytes

```{R}
lapply(PanglaoDBBrainfilter[PanglaoDBBrainfilter$cell.type %in% "Cajal-Retzius cells",]$official.gene.symbol,markervlnplots)
```

Running function for Tancytes

```{R}
lapply(PanglaoDBBrainfilter[PanglaoDBBrainfilter$cell.type %in% "Tanycytes",]$official.gene.symbol,markervlnplots)
```

Glycinergic neurons

```{R}
lapply(PanglaoDBBrainGroupfilter[PanglaoDBBrainGroupfilter$cell.type %in% "Glycinergic neurons",]$official.gene.symbol,markervlnplots)
```

```{R}
lapply(PanglaoDBBrainGroupfilter[PanglaoDBBrainGroupfilter$cell.type %in% "Noradrenergic neurons",]$official.gene.symbol,markervlnplots)
```
```{R}
lapply(PanglaoDBBrainGroupfilter[PanglaoDBBrainGroupfilter$cell.type %in% "Serotonergic neurons",]$official.gene.symbol,markervlnplots)
```

```{R}
lapply(PanglaoDBBrainGroupfilter[PanglaoDBBrainGroupfilter$cell.type %in% "Trigeminal neurons",]$official.gene.symbol,markervlnplots)
```

```{R}
lapply(PanglaoDBBrainGroupfilter[PanglaoDBBrainGroupfilter$cell.type %in% "Pyramidal cells",]$official.gene.symbol,markervlnplots)
```


Running function through lapply for all cell types in the brain in PanglaoDB

```{R}
lapply(unique(PanglaoDBBrain$cell.type), markervlnplots, y=PanglaoDBBrain)



```









Other form of ploting genes across clusters:
```{r}
#RidgePlot(AllMB, features=c("ENO2"))
#RidgePlot(AllMB, features=c("TH"))
#RidgePlot(AllMB, features=c("SLC6A3"))
#RidgePlot(AllMB, features=c("SLC18A2"))
#RidgePlot(AllMB, features=c("SLC17A6"))
#RidgePlot(AllMB, features=c("GAD2"))
```




```{R}
VlnPlot(Batch6_MTG, features = c("CLDN5","TH","P2RY12","GRM4","BCAN"),pt.size = 0,stack = TRUE, flip = TRUE)
```




 Glu_Neurons = 0,1,11,20
 GABA_Neurons = 5,6,9,10,(14),23,24
 Astrocytes = 7,21
 oligo = 3,(4)
 OPCs= 14
 Microglia = 12
 Endothelial = 17, 18
 novel = 2,8,13,15,16,19,20,22,25 - all likely neurons


Assigning cell type identity to clusters: 

```{r}
Batch6_MTG <- RenameIdents(Batch6_MTG, `0` = "Glu_Neurons", `1` = "Glu_Neurons", `2` = "Novel Cluster 2",
                      `3` = "Oligodendrocytes", `4` = "Novel Cluster 4", `5` = "GABA_Neurons",
                      `6` = "GABA Neurons", `7` = "Astrocytes", `8` = "Novel Cluster 4",`9` = "GABA Neurons",
                      `10` = "GABA Neurons", `11` = "Glu_Neurons",`12` = "Microglia",
                      `13` = "Novel Cluster 13",`14` = "GABA Neurons",
                      `15` = "Novel Cluster 15", `16`="Novel Cluster 16", `17`="Endothelial", `18`="Endothelial",`19` = "Novel Cluster 19",
                      `20` = "Novel Cluster 20",`21` = "Astrocytes",
                      `22` = "Novel Cluster 22", `23`="GABA Neurons", `24`="GABA Neurons", `25`="Novel")
```

```{R}
#saveRDS(AllMB,"AllMB_Assigned.rds")
#AllMB=readRDS("AllMB.rds")
```

```{R}

#AllMB <- readRDS("Files/AllMB_Assigned.rds")

```



Now let's plot cells with the assigned celltypes: 
```{r}
DimPlot(Batch6_MTG, label = TRUE, repel = TRUE, pt.size = 0, label.size = 2)

UMAPclusters <- DimPlot(Batch6_MTG, label = TRUE, repel = TRUE, pt.size = 0, label.size = 2.5) + 
        theme(axis.text = element_text(size=8),
              axis.title = element_text(size = 12),
              legend.text = element_text(size = 8),
              title = element_text(size = 12),
              legend.key.size = unit(0.4,"cm"))

ggsave(UMAPclusters, filename = "Files/UMAPclusters_scRNA_seq_Batch6.pdf", device = "pdf", width = 6, height = 4, units = "in")
```

```{r}
DimPlot(Batch6_MTG, label = TRUE, repel = TRUE, pt.size = 1, split.by = "case", label.size = 2,
        ncol = 1) + NoLegend()
```


```{r}
DimPlot(Batch6_MTG, label = TRUE, repel = TRUE, pt.size = 1,reduction = 'umap')
```




